src/ConnectedMode.UnitTests/ServerEventFilteredChannelTests.cs

```diff
@@ -0,0 +1,152 @@
+            MefTestHelpers.CheckTypeCanBeImported<IssueChangedServerEventChannel, IIssueChangedServerEventSourcePublisher>();
+        }
+
+        [TestMethod]
+        public async Task Get_AfterPublish_ReturnsCorrectValue()
+        {
+            var testSubject = CreateTestSubject();
+            var serverEvent = Mock.Of<ITaintServerEvent>();
+
+            testSubject.Publish(serverEvent);
+            var receivedEvent = await testSubject.GetNextEventOrNullAsync();
+
+            receivedEvent.Should().BeSameAs(serverEvent);
+        }
+
+        [TestMethod]
+        public async Task Get_AwaitsForPublishBeforeReturning()
+        {
+            var testSubject = CreateTestSubject();
+            var serverEvent = Mock.Of<ITaintServerEvent>();
+
+            var getTask =  testSubject.GetNextEventOrNullAsync();
+            testSubject.Publish(serverEvent);
+            var receivedEvent = await getTask;
+
+            receivedEvent.Should().BeSameAs(serverEvent);
+        }
+
+        [TestMethod]
+        public async Task Get_ReturnsMultiplePublishedItemsInCorrectOrder()
+        {
+            var testSubject = CreateTestSubject();
+            var serverEvents = Enumerable.Range(0, 5).Select(_ => Mock.Of<ITaintServerEvent>()).ToList();
+            var receivedEvents = new List<ITaintServerEvent>();
+
+            foreach (var serverEvent in serverEvents)
+            {
+                testSubject.Publish(serverEvent);
+            }
+            for (var i = 0; i < serverEvents.Count; i++)
+            {
+                receivedEvents.Add(await testSubject.GetNextEventOrNullAsync());
+            }
+
+            receivedEvents.Should().BeEquivalentTo(serverEvents);
+        }
+
+        [TestMethod]
+        public async Task Get_AlreadyDisposedAndNoMoreItems_ReturnsNull()
+        {
+            var testSubject = CreateTestSubject();
+            var serverEvent = Mock.Of<ITaintServerEvent>();
+
+            testSubject.Publish(serverEvent);
+            testSubject.Dispose();
+            var receivedEvent = await testSubject.GetNextEventOrNullAsync();
+            var receivedEvent2 = await testSubject.GetNextEventOrNullAsync();
+
+            receivedEvent.Should().BeSameAs(serverEvent);
+            receivedEvent2.Should().BeNull();
+        }
+
+        [TestMethod]
+        public async Task Get_DisposedWhileAwaiting_ReturnsNull()
+        {
+            var testSubject = CreateTestSubject();
+
+            var getTask = testSubject.GetNextEventOrNullAsync();
+            testSubject.Dispose();
+            var receivedEvent = await getTask;
+
+            receivedEvent.Should().BeNull();
+        }
+
+        [TestMethod]
+        public void Publish_AlreadyDisposed_ThrowsObjectDisposedException()
+        {
+            var testSubject = CreateTestSubject();
+            var serverEvent = Mock.Of<ITaintServerEvent>();
+
+            testSubject.Publish(serverEvent);
```

Comment:
is the first publish relevant?

Reply:
Yes, to simulate that it also throws even if there are some events in the channel

Reply:
Maybe use TestCases then, `bool isEmptyChannel`